chainntnfs/bitcoindnotify/bitcoind.go

```diff
@@ -796,41 +873,74 @@ func (b *BitcoindNotifier) historicalSpendDetails(
 		default:
 		}
 
-		// First, we'll fetch the block for the current height.
-		blockHash, err := b.chainConn.GetBlockHash(int64(height))
+		// Receive the next block hash from the async queue.
+		blockHash, err := batchRequest[height].Receive()
 		if err != nil {
 			return nil, fmt.Errorf("unable to retrieve hash for "+
 				"block with height %d: %v", height, err)
 		}
-		block, err := b.GetBlock(blockHash)
+
+		blockHashes = append(blockHashes, blockHash)
+	}
+
+	// Now we fetch blocks in interval of 15 to avoid out of memory errors
+	// in case of fetching too many blocks with GetBlocksBatch().
+	const batchSize = 15
+	total := len(blockHashes)
+
+	for i := 0; i < total; i += batchSize {
+		// Ensure we haven't been requested to shut down before
+		// processing next set of blocks.
+		select {
+		case <-b.quit:
+			return nil, chainntnfs.ErrChainNotifierShuttingDown
+		default:
+		}
+
+		start := i
+		end := i + batchSize
+
+		if end > total {
+			end = total
+		}
+
+		blocks, err := b.chainConn.GetBlocksBatch(
+			blockHashes[start:end],
+		)
 		if err != nil {
-			return nil, fmt.Errorf("unable to retrieve block "+
-				"with hash %v: %v", blockHash, err)
+			return nil, err
 		}
 
-		// Then, we'll manually go over every input in every transaction
-		// in it and determine whether it spends the request in
-		// question. If we find one, we'll dispatch the spend details.
-		for _, tx := range block.Transactions {
-			matches, inputIdx, err := spendRequest.MatchesTx(tx)
-			if err != nil {
-				return nil, err
-			}
-			if !matches {
-				continue
-			}
+		// Note:- blockHashes are stored in reverse order
+		// endHeight --> startHeight, so we maintain the same refs
```

Comment:
I mean sure we can reverse the slice and traverse on it, changes would be in +ve sign that's all, something like:

`height:= int(startHeight) + start`
SpendingHeight:    int32(height + j)

Reply:
Yes, i just found it counter intuitive because we are counting backwards. It was just a small nit, nothing to be fixed :+1: .

`SpendingHeight:    int32(startHeight + start + j)` vs `SpendingHeight:    int32(endHeight - start - j)`

Reply:
Again, probably a good idea to refactor this into a method so we have the same reverse loop logic only once in the code. Then the batch size of `15` could also be a parameter there.